pkg/retry/retry.go

```diff
@@ -104,61 +112,80 @@ func (s *Server) handleRetry() http.Handler {
 				return
 			}
 
-			logger.Infow("retrieve deliveries from GitHub", "checkpoint", cursor, "pageSize", len(deliveries))
-
-			// append to the total events counter
-			totalEventCount += len(deliveries)
+			logger.Infow("retrieve deliveries from GitHub", "cursor", cursor, "size", len(deliveries))
 
 			// update the cursor
 			cursor = res.Cursor
 
 			// for each failed delivery, redeliver
 			for i := 0; i < len(deliveries); i++ {
+				// append to the total events counter
+				totalEventCount += 1
+
 				event := deliveries[i]
+
+				// reached the last successful checkpoint, all events equal to and older
+				// than this one have already been processed
+				if lastCheckpoint == strconv.FormatInt(*event.ID, 10) {
+					processed = true
+					break
+				}
+
 				// check payload and see if its been successfully delivered, if so skip over it
 				if *event.StatusCode >= 200 && *event.StatusCode <= 299 {
 					continue
 				}
 
+				// append to the total failure events counter
+				failedEventCount += 1
+
 				logger.Infow("redeliver failed event", "eventID", *event.ID, "guid", *event.GUID)
 
 				if err := s.github.RedeliverEvent(ctx, *event.ID); err != nil {
-					logger.Errorw("failed to call RedeliverEvent",
+					logger.Errorw("failed to call RedeliverEvent, stop processing",
 						"code", http.StatusInternalServerError,
 						"body", errCallingGitHub,
 						"method", "RedeliverEvent",
 						"event", *event,
 						"guid", *event.GUID,
 						"error", err,
+						"totalEventCount", totalEventCount,
+						"failedEventCount", failedEventCount,
 					)
 					w.WriteHeader(http.StatusInternalServerError)
 					fmt.Fprint(w, http.StatusText(http.StatusInternalServerError))
 					return
 				}
-
-				// append to the total failure events counter
-				failedEventCount += 1
 			}
 
-			logger.Infow("write latest checkpoint", "checkpoint", cursor)
+			// every event from newCheckpoint to lastCheckpoint has been processed,
+			// overwrite the checkpoint
+			if processed {
+				logger.Infow("write new checkpoint", "lastCheckpoint", lastCheckpoint, "newCheckpoint", newCheckpoint)
 
-			if err := s.datastore.WriteCheckpointID(ctx, s.checkpointTableID, cursor, now.Format(time.DateTime)); err != nil {
-				logger.Errorw("failed to call WriteCheckpointID",
-					"code", http.StatusInternalServerError,
-					"body", errWriteCheckpoint,
-					"method", "RedeliverEvent",
-					"error", err)
-				w.WriteHeader(http.StatusInternalServerError)
-				fmt.Fprint(w, http.StatusText(http.StatusInternalServerError))
-				return
+				if err := s.datastore.WriteCheckpointID(ctx, s.checkpointTableID,
+					strconv.FormatInt(*newCheckpoint, 10), now.Format(time.DateTime)); err != nil {
```

Comment:
Since we're refactoring this: maybe name these variables before inserting them into the function call?